### PR TITLE
build: introduce 'common' module

### PR DESCRIPTION
**Pull request by david-allison**
## Purpose / Description
I proposed splitting out modules, and 'compat' and 'libAnki' were deemed to be good first targets. We need a ~~'core'~~ 'common' module to handle circular dependencies

See README.md - if I didn't explain it well enough there, it should be added there, rather than the PR description

## Fixes
* Starts https://github.com/ankidroid/Anki-Android/issues/12582

## Approach
* Add Module
* Fixup to meet our JUnit/dependencies
* ~~Move `AdaptionUtil` to the correct place~~
* Extract a couple properties to `TestUtils`

## How Has This Been Tested?
Unit test only

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: core/README.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+## AnkiDroid Core
+
+AnkiDroid core is intended to have no AnkiDroid dependencies, containing utility function, and definitions for core functionality used by all modules
```
Comment (mikehardy) [inline core/README.md:3]: Do you mean no "Android dependencies" (currently says "AnkiDroid dependencies") or is this in fact correct? Just checking

File: core/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    alias(libs.plugins.android.library)
+    alias(libs.plugins.kotlin.android)
+}
+
+android {
+    // this cannot conflict with com.ichi2.anki
+    // but we can define files in 'com.ichi2.anki' inside core
+    // even with this namespace
+    namespace = "com.ichi2.anki.core"
+    compileSdk = 34
```
Comment (mikehardy) [inline core/build.gradle.kts:11]: idle thought - is there a way to use the gradle version catalog to contain these magic numbers? (compileSdk / minSdk / targetSdk / compileOptions compatibility / kotlinOptions jvmTarget)

They are repeated a bunch

**Review by mikehardy — COMMENTED** [review]
seems reasonable - had a hodge-podge of comments, mostly at the "musings" level of importance / vagary (apologies)

File: core/README.md (hunk 1 above)
Comment (david-allison) [reply core/README.md:3]: Core would be the base module of the system, it does not depend on any AnkiDroid modules

File: core/build.gradle.kts (hunk 2 above)
Comment (david-allison) [reply core/build.gradle.kts:11]: Likely, could we extract to an issue?

File: core/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    defaultConfig {
+        minSdk = 23
+
+        testInstrumentationRunner = "androidx.test.runner.AndroidJUnitRunner"
+        consumerProguardFiles("consumer-rules.pro")
+    }
+
+    buildTypes {
+        release {
+            isMinifyEnabled = false
```
[the remark this answers is not in the document]
Comment (david-allison) [reply core/build.gradle.kts:22]: This is the default config for a module. Should this be changed?

[the remark this answers is not in the document]
Comment (david-allison) [reply core/build.gradle.kts:22]: From StackOverflow: https://stackoverflow.com/a/61686668/ (not trustworthy):

> If minifyEnabled is set to true by default and the developer forgets to add proguard rules, it can lead to runtime crashes.

I'll fix this

File: core/src/main/java/com/ichi2/anki/utils/TestUtils.kt [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ *  version.
+ *
+ *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with
+ *  this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.ichi2.anki.utils
```
Comment (lukstbit) [inline core/src/main/java/com/ichi2/anki/utils/TestUtils.kt:14]: I think the package should include the module name. So root _com.ichi2.anki.core_(or common) and add to this the subpackages.

File: core/src/main/java/com/ichi2/anki/utils/TestUtils.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ * isRunningAsUnitTest checks if we are in debug or testing environment by checking if org.junit.Test class
+ * is imported.
+ * https://stackoverflow.com/questions/28550370/how-to-detect-whether-android-app-is-running-ui-test-with-espresso
+ */
+val isRunningAsUnitTest: Boolean
+    get() {
+        try {
+            Class.forName("org.junit.Test")
+        } catch (ignored: ClassNotFoundException) {
+            Timber.d("isRunningAsUnitTest: %b", false)
```
Comment (lukstbit) [inline core/src/main/java/com/ichi2/anki/utils/TestUtils.kt:28]: Isn't Timber an Android library making isRunningAsUnitTest an android specific utility?

File: core/build.gradle.kts (hunk 2 above)
Comment (lukstbit) [reply core/build.gradle.kts:11]: Yes, we can use versions catalog for this, but can also be done at a later time.

File: core/README.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
+## AnkiDroid Core
+
+AnkiDroid core is intended to have no dependencies on other AnkiDroid modules. It contains utility functions, and definitions for core functionality used by all modules
```
Comment (Arthur-Milchior) [inline core/README.md:3]: I first thought you wanted to use git submodules. I think a link to https://developer.android.com/topic/modularization or any other place explaining what a module is in android architecture would be helpful, not every dev are familiar with it.
Maybe even state, for other people like me who know git better than android, that document is not discussing building with submodules, such as when we want to modify  "Anki-Android-Backend".

File: core/README.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+Core is used by `libAnki` (which has no Android dependencies), so dependencies on the Android framework should be in packages named `android`
+
+Core functionality is expected to define interfaces which are initialized in the AnkiDroid Android app
+
+* Core functionality
```
Comment (Arthur-Milchior) [inline core/README.md:9]: It's not directly clear to me what I'm reading.
It looks like a tree. But `com.ichi2.anki.utils` , is a subdirectory of `com.ichi2.anki` and both appear at the same level.

I guess this is the content of this particular module. Then, state it

File: core/README.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  * `com.ichi2.anki` - definitions/interfaces exposing core functionality e.g. `CrashReportService`, `UsageAnalytics` 
+* Utilities which are used throughout the project
+  * `com.ichi2.anki.utils` - utility methods
+  * `com.ichi2.anki.utils.ext` - extension methods
+  * `com.ichi2.anki.utils.android` - utilities with a dependency on Android
```
Comment (Arthur-Milchior) [inline core/README.md:14]: I'm confused. How can we have a package with dependency on android in core if core is supposed not to have dependency to android ?

File: core/README.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+Core is used by `libAnki` (which has no Android dependencies), so dependencies on the Android framework should be in packages named `android`
+
+Core functionality is expected to define interfaces which are initialized in the AnkiDroid Android app
+
+* Core functionality
+  * `com.ichi2.anki` - definitions/interfaces exposing core functionality e.g. `CrashReportService`, `UsageAnalytics` 
+* Utilities which are used throughout the project
+  * `com.ichi2.anki.utils` - utility methods
+  * `com.ichi2.anki.utils.ext` - extension methods
```
Comment (Arthur-Milchior) [inline core/README.md:13]: I'm not against it, but I admit that it's not clear to me why we need to split this way. Unless we maybe do one file by class we extend.

Generally, deciding whether a method is an extension or not is a syntactical choice that I would not expect to be relevant to the packaging choice.


Actually, I suspect it would be worth doing more than a list. And doing sub-sub-scetion and adding one example of the content we can expect to find in each package, to explain show why need it in core

File: core/README.md [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  * `com.ichi2.anki.utils.android` - utilities with a dependency on Android
+
+## Context
+
+It was decided to split out modules in https://github.com/ankidroid/Anki-Android/issues/12582
+
+An ideal module split would be `libAnki` and `AnkiDroid`, both of which require some shared utility
+functions which would have caused circular dependencies
+
+We did not want Android dependencies in `libAnki` once completed
```
Comment (Arthur-Milchior) [inline core/README.md:23]: Maybe I'm just tired, but I admit I find it quite confusing to see a past tense to state what is the long term goal (so future).

I'd rephrase it as:

> ## Context
> 
> This is a work in progress. As discussed in #12582, AnkiDroid team decided to split our codebase in two modules, `libAnki` (the business-logic) and `AnkiDroid` (The part of the code base interacting with Android APIs). 
>
> At the time of writing, this split is not yet done. We expect to do it with the following steps:

**Comment by Arthur-Milchior** [comment]
I admit I'm confused by the argument about circular dependencies.

I would assume that AnkiDroid module depends on LibAnki. I.e., it sends request to the back-end. And libanki does not depends on AnkiDroid. Simple, no circular dependencies, no need for a core.
I assume you already thought about that. So, I got it wrong.
Maybe you mean that no dependency at all should occur. AnkiDroid should not depends on LibAnki at all. In this case, I'd rephrase "no circular dependency" by "no module need to know the other module" or something similar. And also state more explicitly how I'd ensure that, when an action occurs in the UI, the app knows how to update the collection.


In any case, I love that my request of two years ago is looked at.

File: core/src/main/java/com/ichi2/anki/utils/TestUtils.kt (hunk 5 above)
Comment (david-allison) [reply core/src/main/java/com/ichi2/anki/utils/TestUtils.kt:28]: Yes and no - we should document this.

`Timber.d` has no Android dependencies: it loops through the registered loggers (of which there are none).

It doesn't log, but we can define a 'pure java' test logger if desired

The following runs with no issue

```patch
Index: core/src/test/java/com/ichi2/anki/AA.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/core/src/test/java/com/ichi2/anki/AA.kt b/core/src/test/java/com/ichi2/anki/AA.kt
new file mode 100644
--- /dev/null	(date 1717170168330)
+++ b/core/src/test/java/com/ichi2/anki/AA.kt	(date 1717170168330)
@@ -0,0 +1,28 @@
+/*
+ *  Copyright (c) 2024 David Allison <davidallisongithub@gmail.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify it under
+ *  the terms of the GNU General Public License as published by the Free Software
+ *  Foundation; either version 3 of the License, or (at your option) any later
+ *  version.
+ *
+ *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with
+ *  this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.ichi2.anki
+
+import com.ichi2.anki.utils.isRunningAsUnitTest
+import org.hamcrest.MatcherAssert.assertThat
+import org.junit.Test
+
+class AA {
+    @Test
+    fun aa() {
+        assertThat("Is a unit test", isRunningAsUnitTest)
+    }
+}
\ No newline at end of file

```

File: core/README.md (hunk 8 above)
Comment (david-allison) [reply core/README.md:14]: Core has no AnkiDroid depdendencies, it has Android dependencies

**Comment by david-allison** [comment]
Circular dependencies: the module structure should be a digraph. 

I personally want to split out more modules than just `libAnki`, but the issue applies to `libAnki` just as much:

`libAnki` currently has circular dependencies, but shouldn't. We can work on these over time. 

```mermaid
flowchart LR
    subgraph AnkiDroid
        CrashReportService
        deepClonedInto
    end
    subgraph libAnki
        Deck
        DB
    end
    AnkiDroid --> libAnki
    Deck --> deepClonedInto
    DB --> CrashReportService
```

And how 'common' fixes this:

----

```mermaid
flowchart LR
    subgraph AnkiDroid
    end
    subgraph libAnki
        Deck
        DB
    end
    subgraph common
        CrashReportService
        deepClonedInto
    end
    AnkiDroid -- depends on --> libAnki
    Deck --> deepClonedInto
    DB --> CrashReportService
    AnkiDroid -- defines --> CrashReportService
```

File: core/README.md (hunk 9 above)
Comment (david-allison) [reply core/README.md:13]: [`com.ichi2.anki.utils.ext`](https://github.com/ankidroid/Anki-Android/tree/7146af94e83f5347e462261d69b3bfa16c98624b/AnkiDroid/src/main/java/com/ichi2/anki/utils/ext
) has existed for a couple of months. We're currently using one class per file we extend. Brayan introduced it, and I like it from a navigational perspective. 

example: For general `JSONObject` methods, `ext.JSONObject` is a good place to look and an obvious place to add these methods

An additional benefit is that extensions become obvious in the imports list

`import com.ichi2.anki.utils.ext.ifNotZero`

Comment (david-allison) [reply core/README.md:13]: Sub-sub sections: I'm happy with the change, as long as these are on the same level

File: core/build.gradle.kts (hunk 2 above)
Comment (david-allison) [reply core/build.gradle.kts:11]: * https://github.com/ankidroid/Anki-Android/issues/16517

**Comment by AbdelrahmanEsam** [comment]
https://github.com/android/nowinandroid 

@Arthur-Milchior  here is good example from google sample now in android

**Comment by AbdelrahmanEsam** [comment]
@david-allison  I am not yet familiar with  anki code but I can help under your mentoring

**Comment by david-allison** [comment]
@AbdelrahmanEsam this is a pull request

**Comment by AbdelrahmanEsam** [comment]
@david-allison  yes I mean in  general migration plan ....I think you will have a plan to migrate more right ?  to have one core module as interfacing module and make the app modularized by feature and implement the core module in every feature ? ..... or you don't have a plan to do that ?

**Comment by david-allison** [comment]
I'm not sure I understand what you're saying or asking

**Comment by AbdelrahmanEsam** [comment]
@david-allison  will we have modularization plan for our features after the core ?

**Comment by AbdelrahmanEsam** [comment]
@david-allison  
https://galex.dev/posts/advanced-modularization-api-impl-vs-api-di/
here is another way to fix the problem too

**Comment by david-allison** [comment]
This discussion would be better in the issue 

* https://github.com/ankidroid/Anki-Android/issues/12582

But no, no plans to go anywhere near that depth of modules.

I'd expect:

* core
* libAnki
* compat
* wigdets
* app

File: common/README.md [hunk 11]
```diff
@@ -0,0 +1,66 @@
+## AnkiDroid Common
+
+AnkiDroid Common is a [Gradle module](https://developer.android.com/topic/modularization) 
+containing utility functions, and definitions for core functionality used by other modules 
+modules within AnkiDroid. Common should be the base of the AnkiDroid dependency tree.
```
Comment (BrayanDSO) [inline common/README.md:5]: used by other modules modules

**Comment by BrayanDSO** [comment]
As a goal, I believe that we can get rid of the `CrashReportService` dependency in `compat` (easy) and from `libanki` (moderate)

File: common/README.md (hunk 11 above)
Comment (david-allison) [reply common/README.md:5]: Gah... that forced newline got me. Cheers!

**Comment by david-allison** [comment]
Are we likely going to want another cleanup annotation to handle the module cleanups?

Personally, I'd want the benefit from splitting `libAnki` out sooner, rather than after the structure is ideal

File: core/build.gradle.kts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikehardy) [reply core/build.gradle.kts]: yes - the proguard rules (and my mention of "for risk of crashing") is the crux of it - the minifier will prune anything it feels is unreachable, but via reflection and other mechanisms it is possible to rely on / use objects that the minifier can't determine you needed, so it prunes them and you crash. So the process of enabling minification is to flip it on, then run your app like that, and when you crash for whatever reason with a ClassNotFoundException (or kotlin equivalent?) you determine (usually pretty quickly) why (and it's quick because the answer is almost always "oh yeah, we used reflection to load that thing...") and you put a "keepclass" or similar in your proguard rules and go again

**Review by mikehardy — APPROVED** [review]
Just read through it all again and the responses to my earlier musings
Seems good enough to go in and get this started, to me

I see all sorts of the comments and discussions going on (fantastic), so just leaving a green check as a vote, I'm personally satisfied at least

**Comment by david-allison** [comment]
I'm not in a rush to merge this. Happy to wait until everyone's happy

**Review by BrayanDSO — APPROVED** [review]
also LGTM

**Comment by david-allison** [comment]
@Arthur-Milchior Awaiting review

File: common/README.md [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+Definitions/interfaces exposing core functionality e.g. `CrashReportService`, `UsageAnalytics`
+
+These are to be initialized higher up the dependency tree, typically in `AnkiDroid`
+
+### `com.ichi2.anki.common.utils`
+
+Utility classes and methods without an Android dependency
+
+
```
Comment (Arthur-Milchior) [inline common/README.md:24]: I'd go consistently with a single line before a new subsection title. Not blocking

File: common/README.md [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+Extension methods, universally applicable to the classes they extend
+
+Examples:
+
+* `Int.kt` - `ifNotZero`
+* `InputStream.kt` - `convertToString`
+
+### `com.ichi2.anki.common.utils.android` 
+
+Utilities with a dependency on Android
```
Comment (Arthur-Milchior) [inline common/README.md:36]: I am sorry for I'm still confused. If libanki depends on commons which depends on android, it seems that libanki indirectly depends on android. 
Libanki may not actually import any android package, and that would be easy to check using a linter for example.   Still, it means that when building libanki that would include any android dependencies that were imported by commons package.

I would have expected any such method to be in the AnkiDroid module.

**Review by Arthur-Milchior — APPROVED** [review]
I am confused, but the documentation can be improved later. I see nothing that would require blocking merging the PR.
